### PR TITLE
add dataclass support on apply_to_collection

### PR DESCRIPTION
**Pull request by dalek-who**
## What does this PR do?

issue：#7878
Add dataclass support on apply_to_collection, not only on move_batch_to_device

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/apply_func.py [hunk 1]
```diff
@@ -110,6 +111,14 @@ def apply_to_collection(
                 out.append(v)
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if dataclasses.is_dataclass(data) and not isinstance(data, type):
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:114]: why the `and not isinstance(data, type)` check?

File: pytorch_lightning/utilities/apply_func.py [hunk 2]
```diff
@@ -110,6 +111,14 @@ def apply_to_collection(
...
+        out = []
+        for field in data.__dataclass_fields__:
+            v = apply_to_collection(getattr(data, field), dtype, function, *args, wrong_dtype=wrong_dtype, **kwargs)
+            if include_none or v is not None:
+                out.append((field, v))
+        return elem_type(**OrderedDict(out))
```
Comment (carmocca) [inline pytorch_lightning/utilities/apply_func.py:120]: Shouldn't need `OrderedDict` in this case as the values are mapped by keyword

File: CHANGELOG.md [hunk 3]
```diff
@@ -9,6 +9,9 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 
 ### Added
 
+- Add `dataclass` support for `pytorch_lightning.utilities.apply_func` ([#7935](https://github.com/PyTorchLightning/pytorch-lightning/pull/7935))
```
Comment (awaelchli) [inline CHANGELOG.md:12]:
```suggestion
- Add `dataclass` support for `pytorch_lightning.utilities.apply_to_collection` ([#7935](https://github.com/PyTorchLightning/pytorch-lightning/pull/7935))
```
I think users will recognize `apply_to_collection` better when reading the changelog so I suggest this

File: tests/utilities/test_apply_func.py [hunk 4]
```diff
@@ -32,6 +45,12 @@ def test_recursive_application_to_collection():
         'e': np.array([10.]),  # numpy array
         'f': 'this_is_a_dummy_str',  # string
         'g': 12.,  # number
+        'h': Feature(input_ids=torch.tensor([1., 2., 3.]), segment_ids=np.array([4., 5., 6.])),  # dataclass
+        "i": ModelExample(
```
Comment (awaelchli) [inline tests/utilities/test_apply_func.py:49]: could you standardize the string quotation marks here and below, sorry for the useless comment, our auto formatter does not yet do that automatically ^^

**Review by awaelchli — APPROVED** [review]
nice addition
dataclasses reduce boilerplate code, so does Lighting. We should make best friends!  <3

File: pytorch_lightning/utilities/apply_func.py (hunk 1 above)
Comment (dalek-who) [reply pytorch_lightning/utilities/apply_func.py:114]: This is the official recommended way:
> https://docs.python.org/3/library/dataclasses.html#module-level-decorators-classes-and-functions
> 
> dataclasses.is_dataclass(class_or_instance)
> Return True if its parameter is a dataclass or an instance of one, otherwise return False.
> If you need to know if a class is an instance of a dataclass (and not a dataclass itself), then add a further check for not isinstance(obj, type):
> ```python
> def is_dataclass_instance(obj):
>     return is_dataclass(obj) and not isinstance(obj, type)
> ```

**Review by justusschock — APPROVED** [review]

**Review by tchaton — APPROVED** [review]
LGTM ! Nice addition.

File: pytorch_lightning/utilities/apply_func.py (hunk 1 above)
Comment (carmocca) [reply pytorch_lightning/utilities/apply_func.py:114]: I think it would be ncie for you to copy that function and use it, just as we have for `_is_namedtuple`

File: pytorch_lightning/utilities/apply_func.py [hunk 5]
```diff
@@ -102,6 +108,7 @@ def apply_to_collection(
 
     is_namedtuple = _is_namedtuple(data)
     is_sequence = isinstance(data, Sequence) and not isinstance(data, str)
+    is_dataclass = _is_dataclass_instance(data)
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:111]:
```suggestion
```

File: pytorch_lightning/utilities/apply_func.py [hunk 6]
```diff
@@ -110,6 +117,14 @@ def apply_to_collection(
                 out.append(v)
         return elem_type(*out) if is_namedtuple else elem_type(out)
 
+    if is_dataclass:
```
Comment (Borda) [inline pytorch_lightning/utilities/apply_func.py:120]:
```suggestion
    if _is_dataclass_instance(data):
```

**Review by Borda — APPROVED** [review]

**Review by carmocca — APPROVED** [review]
